diff --git a/Cargo.toml b/Cargo.toml
index 37ee5313bf1..a4d29cfa812 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -36,9 +36,9 @@ app_units = "0.7"
 arboard = "3"
 argon2 = { version = "0.5", features = ["alloc"] }
 arrayvec = "0.7"
-async-tungstenite = { version = "0.34", features = ["tokio-rustls-webpki-roots"] }
+async-tungstenite = { version = "0.34", default-features = false, features = ["tokio-runtime", "tokio-rustls-manual-roots"] }
 atomic_refcell = "0.1.14"
-aws-lc-rs = { version = "1.16", default-features = false, features = ["aws-lc-sys"] }
+aws-lc-rs = { version = "1.16", default-features = false }
 backtrace = "0.3"
 base64 = "0.22.1"
 base64ct = { version = "1.8", features = ["alloc"] }
@@ -160,7 +160,7 @@ regex = "1.12"
 resvg = "0.45.0"
 rsa = { version = "0.9.10", features = ["sha1", "sha2"] }
 rustc-hash = "2.1.2"
-rustls = { version = "0.23", default-features = false, features = ["logging", "std", "tls12"] }
+rustls = { version = "0.23", default-features = false, features = ["logging", "std", "tls12", "ring"] }
 rustls-pki-types = "1.14"
 rustls-platform-verifier = "0.7.0"
 sea-query = { version = "=1.0.0-rc.31", default-features = false, features = ["backend-sqlite", "derive"] }
diff --git a/components/allocator/Cargo.toml b/components/allocator/Cargo.toml
index 8af83fb658e..2e8a1251d8a 100644
--- a/components/allocator/Cargo.toml
+++ b/components/allocator/Cargo.toml
@@ -21,7 +21,7 @@ backtrace = { workspace = true, optional = true }
 log = { workspace = true, optional = true }
 rustc-hash = { workspace = true, optional = true }
 
-[target.'cfg(not(any(windows, target_env = "ohos")))'.dependencies]
+[target.'cfg(not(any(windows, target_os = "freebsd", target_env = "ohos")))'.dependencies]
 libc = { workspace = true, optional = true }
 tikv-jemalloc-sys = { workspace = true }
 tikv-jemallocator = { workspace = true }
@@ -31,3 +31,6 @@ windows-sys = { workspace = true, features = ["Win32_System_Memory"] }
 
 [target.'cfg(target_env = "ohos")'.dependencies]
 libc = { workspace = true }
+
+[target.'cfg(target_os = "freebsd")'.dependencies]
+libc = { workspace = true }
diff --git a/components/allocator/lib.rs b/components/allocator/lib.rs
index a64ff560cc6..100edc38576 100644
--- a/components/allocator/lib.rs
+++ b/components/allocator/lib.rs
@@ -53,7 +53,7 @@ pub static enclosing_size: Option<EnclosingSizeFn> = Some(crate::enclosing_size_
 #[cfg(not(feature = "allocation-tracking"))]
 pub static enclosing_size: Option<EnclosingSizeFn> = None;
 
-#[cfg(not(any(windows, feature = "use-system-allocator", target_env = "ohos")))]
+#[cfg(not(any(windows, target_os = "freebsd", feature = "use-system-allocator", target_env = "ohos")))]
 mod platform {
     use std::os::raw::c_void;
 
@@ -77,10 +77,7 @@ mod platform {
     }
 }
 
-#[cfg(all(
-    not(windows),
-    any(feature = "use-system-allocator", target_env = "ohos")
-))]
+#[cfg(all(not(windows), any(feature = "use-system-allocator", target_os = "freebsd", target_env = "ohos")))]
 mod platform {
     pub use std::alloc::System as Allocator;
     use std::os::raw::c_void;
diff --git a/components/net/Cargo.toml b/components/net/Cargo.toml
index 170a0e9b15f..97e9908ab3e 100644
--- a/components/net/Cargo.toml
+++ b/components/net/Cargo.toml
@@ -63,7 +63,7 @@ quick_cache = { version = "0.6.18", default-features = false, features = ["ahash
 regex = { workspace = true }
 resvg = { workspace = true }
 rustc-hash = { workspace = true }
-rustls = { workspace = true, features = ["aws-lc-rs"] }
+rustls = { workspace = true, features = ["ring"] }
 rustls-pki-types = { workspace = true }
 rustls-platform-verifier = { workspace = true }
 serde = { workspace = true }
@@ -96,7 +96,7 @@ hyper-util = { workspace = true, features = ["server-graceful"] }
 # This must always be a path dependency, we don't want to add a dependency on a
 # crates.io version of ourselves - we want to enable a feature of ourselves.
 net = { package = "servo-net", path = ".", features = ["test-util"] }
-rustls = { workspace = true, features = ["aws-lc-rs"] }
+rustls = { workspace = true, features = ["ring"] }
 
 [[test]]
 name = "main"
diff --git a/components/profile/Cargo.toml b/components/profile/Cargo.toml
index 7f8d3847524..a6b4fef44da 100644
--- a/components/profile/Cargo.toml
+++ b/components/profile/Cargo.toml
@@ -32,5 +32,5 @@ regex = { workspace = true }
 [target.'cfg(not(target_os = "windows"))'.dependencies]
 libc = { workspace = true }
 
-[target.'cfg(not(any(target_os = "windows", target_env = "ohos")))'.dependencies]
+[target.'cfg(not(any(target_os = "windows", target_os = "freebsd", target_env = "ohos")))'.dependencies]
 tikv-jemalloc-sys = { workspace = true }
diff --git a/components/profile/system_reporter.rs b/components/profile/system_reporter.rs
index 97996d3f532..48353ab3678 100644
--- a/components/profile/system_reporter.rs
+++ b/components/profile/system_reporter.rs
@@ -2,16 +2,16 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
-#[cfg(not(any(target_os = "windows", target_env = "ohos")))]
+#[cfg(not(any(target_os = "windows", target_os = "freebsd", target_env = "ohos")))]
 use std::ffi::CString;
-#[cfg(not(any(target_os = "windows", target_env = "ohos")))]
+#[cfg(not(any(target_os = "windows", target_os = "freebsd", target_env = "ohos")))]
 use std::mem::size_of;
-#[cfg(not(any(target_os = "windows", target_env = "ohos")))]
+#[cfg(not(any(target_os = "windows", target_os = "freebsd", target_env = "ohos")))]
 use std::ptr::null_mut;
 
 #[cfg(all(target_os = "linux", target_env = "gnu"))]
 use libc::c_int;
-#[cfg(not(any(target_os = "windows", target_env = "ohos")))]
+#[cfg(not(any(target_os = "windows", target_os = "freebsd", target_env = "ohos")))]
 use libc::{c_void, size_t};
 use profile_traits::mem::{ProcessReports, Report, ReportKind, ReporterRequest};
 use profile_traits::path;
@@ -164,10 +164,10 @@ fn system_heap_info() -> SystemHeapInfo {
     }
 }
 
-#[cfg(not(any(target_os = "windows", target_env = "ohos")))]
+#[cfg(not(any(target_os = "windows", target_os = "freebsd", target_env = "ohos")))]
 use tikv_jemalloc_sys::mallctl;
 
-#[cfg(not(any(target_os = "windows", target_env = "ohos")))]
+#[cfg(not(any(target_os = "windows", target_os = "freebsd", target_env = "ohos")))]
 fn jemalloc_stat(value_name: &str) -> Option<usize> {
     // Before we request the measurement of interest, we first send an "epoch"
     // request. Without that jemalloc gives cached statistics(!) which can be
@@ -214,7 +214,7 @@ fn jemalloc_stat(value_name: &str) -> Option<usize> {
     Some(value as usize)
 }
 
-#[cfg(any(target_os = "windows", target_env = "ohos"))]
+#[cfg(any(target_os = "windows", target_os = "freebsd", target_env = "ohos"))]
 fn jemalloc_stat(_value_name: &str) -> Option<usize> {
     None
 }
diff --git a/components/script/Cargo.toml b/components/script/Cargo.toml
index 9c1c72a2cd2..cfe7c35458a 100644
--- a/components/script/Cargo.toml
+++ b/components/script/Cargo.toml
@@ -48,7 +48,6 @@ app_units = { workspace = true }
 argon2 = { workspace = true }
 arrayvec = { workspace = true }
 atomic_refcell = { workspace = true }
-aws-lc-rs = { workspace = true }
 backtrace = { workspace = true }
 base64 = { workspace = true }
 base64ct = { workspace = true }
diff --git a/components/shared/paint/rendering_context.rs b/components/shared/paint/rendering_context.rs
index 0bd82ae05f7..2b4f746ff2f 100644
--- a/components/shared/paint/rendering_context.rs
+++ b/components/shared/paint/rendering_context.rs
@@ -753,7 +753,7 @@ pub struct OffscreenRenderingContext {
     framebuffer: RefCell<Framebuffer>,
 }
 
-type RenderToParentCallback = Box<dyn Fn(&glow::Context, Rect<i32>) + Send + Sync>;
+type RenderToParentCallback = Box<dyn Fn(&glow::Context, Rect<i32>, u32) + Send + Sync>;
 
 impl OffscreenRenderingContext {
     fn new(parent_context: Rc<WindowRenderingContext>, size: PhysicalSize<u32>) -> Self {
@@ -778,16 +778,16 @@ impl OffscreenRenderingContext {
         // Don't accept a `None` context for the source framebuffer.
         let front_framebuffer_id =
             NonZeroU32::new(self.framebuffer.borrow().framebuffer_id).map(NativeFramebuffer)?;
-        let parent_context_framebuffer_id = self.parent_context.surfman_context.framebuffer();
         let size = self.size.get();
         let size = Size2D::new(size.width as i32, size.height as i32);
-        Some(Box::new(move |gl, target_rect| {
+        Some(Box::new(move |gl, target_rect, target_fbo| {
+            let target_framebuffer_id = NonZeroU32::new(target_fbo).map(NativeFramebuffer);
             Self::blit_framebuffer(
                 gl,
                 Rect::new(Point2D::origin(), size.to_i32()),
                 front_framebuffer_id,
                 target_rect,
-                parent_context_framebuffer_id,
+                target_framebuffer_id,
             );
         }))
     }
diff --git a/ports/servoshell/Cargo.toml b/ports/servoshell/Cargo.toml
index 9502f197938..0df3daa1540 100644
--- a/ports/servoshell/Cargo.toml
+++ b/ports/servoshell/Cargo.toml
@@ -76,7 +76,7 @@ libc = { workspace = true }
 log = { workspace = true }
 mime_guess = { workspace = true }
 raw-window-handle = { workspace = true }
-rustls = { workspace = true, features = ["aws-lc-rs"] }
+rustls = { workspace = true, features = ["ring"] }
 servo = { workspace = true, features = ["background_hang_monitor", "bluetooth", "testbinding"], default-features = false }
 tokio = { workspace = true }
 tracing = { workspace = true, optional = true }
